src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -254,29 +351,44 @@ static void CheckForExperimentStatus(ITextView textView, in CompletionOptions op
             {
                 AsyncCompletionLogger.LogExpanderUsage();
 
+                // It's possible we didn't provide expanded items at the beginning of completion session because it was slow even if the feature is enabled.
+                // ExpandedItemsTask would be available in this case, so we just need to return its result.
+                // We are doing this in a spinning fashion because the lifetime of expandedItemsTask is associated with the completion session, we don't
+                // want to cancel that task when this "GetExpandedCompletionContextAsync" operation is being canceled, so we will still be able to retrieve
+                // the results later.
+                if (session.Properties.TryGetProperty<Task<(AsyncCompletionData.CompletionContext, CompletionList)>>(ExpandedItemsTask, out var expandedItemsTask))
+                {
+                    // Make sure the task is removed when returning expanded items,
+                    // so duplicated items won't be added in subsequent list updates.
+                    session.Properties.RemoveProperty(ExpandedItemsTask);
+                    var (expandedContext, expandedCompletionList) = await expandedItemsTask.ConfigureAwait(false);
+                    AddPropertiesToSession(session, expandedCompletionList, triggerLocation: null);
+                    return expandedContext;
```

Comment:
i'm not really sure i get why we return the expandedContext but adding the expactedComlpetionList as a property to the session...

Reply:
taht this is a Get method that mutates properties is a bit oogy.  can the data just flow outwards, and someone else can be responsible for this?

Reply:
>taht this is a Get method that mutates properties is a bit oogy. can the data just flow outwards, and someone else can be responsible for this?

+1 from me

>i'm not really sure i get why we return the expandedContext but adding the expactedComlpetionList as a property to the session...

This was confusing to me as well - but I think the expanded completion list is not actually being added to the properties bag here, right?  Its just taking certain properties from the list and adding them.  It might be good to explicitly pass in what you're adding to the properties bag?